### PR TITLE
Add wall time and error type filters in Slack notification templates

### PR DESCRIPTION
**Pull request by Yaliang**

### CODE REVIEW

File: twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class AbsentOrPattern
+{
+    private final Optional<Boolean> missed;
```
Comment (luohao) [inline twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java]: `valueWhenAbsent` and `getValueWhenAbsent()` might be a better name.

File: twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    private final Optional<Pattern> regex;
+
+    @JsonCreator
+    public AbsentOrPattern(
+            @JsonProperty("missed") Optional<Boolean> missed,
+            @JsonProperty("regex") Optional<Pattern> regex)
+    {
+        checkArgument(missed.isPresent() || regex.isPresent(), "missed and regex are both absent");
```
Comment (luohao) [inline twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java]: Can you add a little bit more comments about the meanings of different combinations to help people understand the code? Not sure my understanding in the table below is correct.

| isPresent() | valueWhenAbsent | regex |               |   
|-------------|---------|-------|---------------|
| 0           | false   | false | Invalid State | 
| 1           | false   | true  | if field is present, match pattern; if field is absent, return false | 
| 2           | true    | false | if field is present, return true; if field is absent, return `valueWhenAbsent ` | 
| 3           | true    | true  | if field is present, match pattern; if field is absent, return `valueWhenAbsent ` |

File: twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/SlackNotificationTemplate.java [hunk 3]
```diff
@@ -15,41 +15,74 @@
...
+        return checkOptionalFields(Optional.of(regex), Optional.empty(), key, fields);
+    }
+
+    private boolean checkOptionalFields(Optional<Pattern> regex, Optional<Boolean> whenFieldAbsent, String key, Map<String, Optional<String>> fields)
+    {
+        Optional<String> field = fields.getOrDefault(key, Optional.empty());
+        if (!field.isPresent() && whenFieldAbsent.isPresent()) {
+            return whenFieldAbsent.get();
+        }
+        return regex.map(r -> field.isPresent() && r.matcher(field.get()).matches()).orElse(true);
```
Comment (luohao) [inline twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/SlackNotificationTemplate.java]: Will there be cases you need to return `false` when `regex` is empty?

Comment (Yaliang) [reply twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/SlackNotificationTemplate.java]: We can use some regex like "a^" to match nothing in that case.

File: twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java (hunk 2 above)
Comment (Yaliang) [reply twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java]: Yeah, you're right. It's a little bit too complex to put all logic here. I can move the logic to `AbsentOrPattern` class. Added comments with an example on the head of the class. Also added a test class to cover most cases.

File: twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java (hunk 1 above)
Comment (Yaliang) [reply twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java]: Does `isAbsent` make sense here? This class should be the definition of the absent-able pattern itself. `valueWhenAbsent` seems to provide a default value of the field when the field is absent.

Comment (Yaliang) [reply twitter-eventlistener-plugin/src/main/java/com/twitter/presto/plugin/eventlistener/slack/AbsentOrPattern.java]: Saying an example:
```
"principal": {
  "isAbsent": true,
  "regex": "user.*"
}
```
is defining a pattern that `principal is absent or principal matches regex`.

**Comment by luohao** [comment]
LGTM.

Merged, thanks!

**Pull request merged**